xayn-ai/src/embedding/utils.rs

```diff
@@ -57,23 +57,31 @@ where
 ///
 /// # Panics
 /// Panics if the vectors don't consist solely of real values or their shapes don't match.
+///
+/// Only use this with [`TrustedLen`] iterators, otherwise indexing may panic. The bound can't be
+/// named currently, because the trait is nightly-gated. [`ExactSizeIterator`] isn't a feasible and
+/// trusted replacement, for example it isn't implemented for [`Chain`]ed iterators.
+///
+/// [`TrustedLen`]: std::iter::TrustedLen
+/// [`ExactSizeIterator`]: std::iter::ExactSizeIterator
+/// [`Chain`]: std::iter::Chain
 #[allow(dead_code)]
-pub fn pairwise_cosine_similarity<'a, I, S>(iter: I) -> Array2<f32>
+pub fn pairwise_cosine_similarity<I, S>(iter: I) -> Array2<f32>
```

Comment:
Seems that we have lot's of complexity for a minor optimization. 
Can't just we do something like:
```rust
iter.clone().enumerate().cartesian_product(iter.enumerate())
  .filter(|((i, _), (j, _))| i > j)
  .map(/* similarity here */)
  .collect()
``` 
?
We are already iterating more then necessary by doing `skip`  anyway.
The other option would be to take a `Vec<ArrayBase>` or a ref to it, so we can allocate the correct amount and iterate only the needed times.

Reply:
i just timed it, the cartesian product approach is about 30% faster and the vector approach is about 30% slower than the skip approach, so i'm going to change it to the former.